src/Microsoft.AspNetCore.Authentication.Cookies/CookieAuthenticationHandler.cs

```diff
@@ -311,14 +311,24 @@ private CookieOptions BuildCookieOptions()
             Response.Headers[HeaderNames.CacheControl] = HeaderValueNoCache;
             Response.Headers[HeaderNames.Pragma] = HeaderValueNoCache;
             Response.Headers[HeaderNames.Expires] = HeaderValueMinusOne;
+
             if (shouldRedirectToReturnUrl && Response.StatusCode == 200)
             {
+                CookieRedirectContext redirectContext = null;
+                
                 var query = Request.Query;
                 var redirectUri = query[Options.ReturnUrlParameter];
-                if (!StringValues.IsNullOrEmpty(redirectUri)
-                    && IsHostRelative(redirectUri))
+                if (!StringValues.IsNullOrEmpty(redirectUri) && IsHostRelative(redirectUri))
```

Comment:
IMO, the redirect_uri stored in the properties should be preferred to the return_url parameter (when it's present).